Feat : Sophia & ForeachAdaLolmo Optimizer

namgyu-youn opened the pull request:
## Description
This PR is rebased (reviewed) version of #51
- close #49 
- close #35 

## Change Log
- Feat : Sophia & AdaLolmo Optimizer
- Update convention rule (`--diff` -> `--fix`)
- Refactor codebase for convention rule

namgyu-youn commented:
@ClashLuke Is anyone here..? 👋

namgyu-youn commented:
@ClashLuke Hello?? I am waiting for you..

ClashLuke commented:
Hi namgyu-youn,

I've been waiting for you to fix the issues detailed in the review of #51. This PR is 95% unchanged and doesn't address any of the feedback. Please fix those issues instead of marking conversations as resolved or opening multiple PRs with essentially identical code.

Removing your previous comment just to ping me again is not appropriate GitHub etiquette. If you genuinely want to contribute, please address the actual feedback provided.

namgyu-youn commented:
Sorry, I found some unfixed issues. Yes, that was my fault; (it occurred while rebasing.) I will fix it immediately.

namgyu-youn commented:
@ClashLuke I am sorry to ping you again, but I want to carefully emphasize that **this PR addresses the feedback**. Please consider the review again.. I wrote more explain for that below.


![image](https://github.com/user-attachments/assets/a3a84be8-7064-43b6-b885-8cf6755fc616)
And `__init__.py Ln 307`, `donom` is fixed for the fallback.

---

![image](https://github.com/user-attachments/assets/38679a8a-46ca-4f9a-874f-8dbcf42195d0)
AdaLolmo's chainable API was placed inapposite in `__init__.py`, so it is moved to `chainable.py Ln 327`

---

![image](https://github.com/user-attachments/assets/7cbbb6b6-464a-4e2b-a52c-19aa624b7907)
It is fixed by the fallback methods in `chainable.py Ln 506`

---

![image](https://github.com/user-attachments/assets/044ae017-cd62-460e-8187-d7ba1cf70ef6)
`u_rms` is fixed with more simple fallback methods. - `chainable.py Ln 503`

ClashLuke commented:
I've done all the ruff reformatting changes on my end, so we can reduce the scope of the PR a bit. Thank you for taking initiative.

ClashLuke commented on heavyball/chainable.py:
```diff
@@ -204,6 +204,410 @@ def update_by_laprop(group, update, grad, param, exp_avg, exp_avg_sq):
+            next_hessian_update[0] = hessian_step[0] + k
+
+            # Hessian approximation using squared gradient
+            h = (grad.detach() * grad.detach())
```
Is this correct? That's the same as Adam

namgyu-youn replied:
The difference against Adam is `update_freq`. In the paper, Sophia uses `t mod k` for the update cycle.

ClashLuke commented on heavyball/chainable.py:
```diff
@@ -204,6 +204,410 @@ def update_by_laprop(group, update, grad, param, exp_avg, exp_avg_sq):
+    copy_stochastic_(update, u_norm)
+
+    # Store the state
+    param.optimizer_state = state
```
the state is managed through general guard. do not modify it like this

namgyu-youn replied:
Thanks. i will remove it.

ClashLuke commented on heavyball/chainable.py:
```diff
@@ -204,6 +204,410 @@ def update_by_laprop(group, update, grad, param, exp_avg, exp_avg_sq):
+        # Compute Hessian-vector product
+        with torch.enable_grad():
+            p = param.detach().requires_grad_()
+            g = torch.autograd.grad(torch.sum(grad * v), p, create_graph=True)[0]
```
This will not work. However, there's already `hvp` code in `utils` that you can hook into.

namgyu-youn replied:
Thanks. Then, may i fix it using `modify_closure`, `StatefulOptimizer` in `utils.py`?

ClashLuke reviewed (changes requested):
There are many bugs in the code and the comments from the previous PR remain unresolved

namgyu-youn commented on heavyball/__init__.py:
```diff
@@ -254,6 +256,269 @@ def __init__(self, params, lr=0.001, beta=0.9, weight_decay=0.0, preconditioner_
+                         C.zero_guard("diag_hessian"),
+                         C.scale_by_sophia)
+
+    def _step(self, group):
```
Sophia updates Hessian metrix in k-steps, and `_steps` is needed for that Second-gradient logic.

If it is removed, it would work like normal First-gradient like Adam, AdamW.

namgyu-youn commented on heavyball/__init__.py:
```diff
@@ -254,6 +256,269 @@ def __init__(self, params, lr=0.001, beta=0.9, weight_decay=0.0, preconditioner_
+        defaults.pop("self")
+        params = defaults.pop("params")
+        super().__init__(params, defaults, foreach, gradient_clipping, update_clipping, palm,
+                         C.update_by_adalomo)
```
Thanks! :)

namgyu-youn commented on heavyball/__init__.py:
```diff
@@ -254,6 +256,269 @@ def __init__(self, params, lr=0.001, beta=0.9, weight_decay=0.0, preconditioner_
+                         C.update_by_adalomo)
+
+
+class SophiaH(torch.optim.Optimizer):
```
If above is `SophiaH`, it doesn't work same like `ForeachSophiaH`. For more explain,
- `ForeachSophiaH` : `C.BaseOpt` -> `ChainOpt` -> `torch.optim.Optimizer` (Chain based API)
- `Sophia` : Inherited by `torch.optim.Optimzer` directly

namgyu-youn commented on heavyball/__init__.py:
```diff
@@ -254,6 +256,269 @@ def __init__(self, params, lr=0.001, beta=0.9, weight_decay=0.0, preconditioner_
+                       gamma=gamma, update_freq=update_freq)
+        super().__init__(params, defaults)
+
+    def step(self, closure=None):
```
But i think this method should be baseline for Sophia implementation. As sure chain methods could be used on this, but i'm not sure that would be better.

namgyu-youn commented on heavyball/__init__.py:
```diff
@@ -293,11 +558,16 @@ class ForeachCachedNewtonPSGD(ForeachCachedPSGDKron):
 Muon = ForeachMuon
 SignLaProp = ForeachSignLaProp
+SophiaH = SophiaH
+SophiaG = SophiaH
```
Because i was planning for implementation after this PR. If you mind it, i would markdown it or remove that.

namgyu-youn commented on heavyball/chainable.py:
```diff
@@ -204,6 +204,410 @@ def update_by_laprop(group, update, grad, param, exp_avg, exp_avg_sq):
+    # Original shape for reshaping if needed
+    original_shape = update.shape
+
+    if len(original_shape) <= 1:
```
Maybe common logic could be seperated methods. If you like that, i would follow your decision.

namgyu-youn commented on heavyball/chainable.py:
```diff
@@ -204,6 +204,410 @@ def update_by_laprop(group, update, grad, param, exp_avg, exp_avg_sq):
+        h32 = utils.promote(h)
+
+        # Compute denominator with safeguard
+        denom = torch.maximum(gamma * h32, torch.tensor(eps, device=h32.device, dtype=h32.dtype))
```
I missed it, thanks!

namgyu-youn commented on heavyball/chainable.py:
```diff
@@ -204,6 +204,410 @@ def update_by_laprop(group, update, grad, param, exp_avg, exp_avg_sq):
+        # Apply update
+        p_data = utils.promote(p.data)
+        p_data.add_(scaled, alpha=-lr)
+        utils.copy_stochastic_(p.data, p_data)
```
Yes it should be wrapped using `@decorator_knowngood`. Decorator always makes me hard.. I will fix it!

namgyu-youn commented on heavyball/__init__.py:
```diff
@@ -254,6 +256,269 @@ def __init__(self, params, lr=0.001, beta=0.9, weight_decay=0.0, preconditioner_
+                         C.scale_by_sophia)
+
+
+    def estimate_hessian_g(self, state, group, update, grad, param):
```
Yes that is dead code... I missed it :confused:. I will remove it

namgyu-youn commented:
> There are many bugs in the code and the comments from the previous PR remain unresolved

I reopened them today. I thought most of the issue was resolved using your review methods (fallback). But in my thought, being reviewed again (for the review) would be necessary. So i reopened them (As sure, they are already resolved).

namgyu-youn commented:
@ClashLuke This ping is just for notification. Please review the last comments and **resolve the conversation** if you want.

namgyu-youn commented:
@ClashLuke Why are you ignoring my PR? I updated fixed many request in [1b85b9f](https://github.com/HomebrewML/HeavyBall/pull/52/commits/1b85b9f9439d22898dafd3c7fb4952832c2a16ab), and requested your review again. Have you ever checked my comment?

ClashLuke commented:
@namgyu-youn I'm closing this PR and blocking you from our repository. Despite multiple reviews, you've submitted essentially identical non-functional code across three PRs, falsely claimed to fix issues, and displayed disruptive communication practices.

Note to other maintainers: this account has exhibited this pattern across repositories.
